core/utils/semaphore.c

```diff
@@ -0,0 +1,116 @@
+/* Semaphore utility for reactor C. */
+
+/*************
+Copyright (c) 2021, The University of Texas at Dallas.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***************/
+
+/** 
+ * Semaphore utility for reactor C.
+ *  
+ * @author{Soroush Bateni <soroush@utdallas.edu>}
+ */
+
+#include "semaphore.h"
+
+/**
+ * @brief Create a new semaphore.
+ * 
+ * @param count The count to start with.
+ * @return semaphore_t* Can be NULL on error.
+ */
+semaphore_t* lf_semaphore_new(int count) {
+    semaphore_t* semaphore = (semaphore_t*)malloc(sizeof(semaphore_t));
+    lf_cond_init(&semaphore->cond);
+    lf_mutex_init(&semaphore->mutex);
+    semaphore->count = count;
+    return semaphore;
+}
+
+/**
+ * @brief Release the 'semaphore' and deduct 'i' from its count.
+ * 
+ * @param semaphore Instance of a semaphore
+ * @param i The count to deduct
+ * @return int -1 on error. 0 on success.
+ */
+int lf_semaphore_release(semaphore_t* semaphore, int i) {
+    if (semaphore == NULL) {
```

Comment:
You're right about the use case. But the return value is not checked in the schedulers for the sake of performance. I still think that the correct C API design for the semaphore is for it to return a status code. This semaphore can be useful in other parts of reactor-c too.

Reply:
I do not disagree with your general point. What I suggested was not very safe. But there were reasons why I thought an `assert` might be a fair alternative here:
* If performance matters in the schedulers, then it may matter here
* This is different from functions like `malloc` (which actually might fail, through no fault of the caller) in that if this check fails, then it was because of an illegal argument -- the caller shot themselves in the foot. If the caller is not responsible enough to ensure that the pointer is not NULL, then why should the caller be responsible enough to check the return code?
* In many cases, not doing the check when the semaphore is NULL would result in a segfault (I think?), which would crash the program and allow the user to know that something is terribly wrong. In this sense, this behavior could even be more idiot-proof in some ways than relying on the user to check the return code.

Reply:
Good point. Let's use your simpler API design.